### PR TITLE
Fixes #36849 - Run GHA on Ruby 3.0

### PR DESCRIPTION
**Pull request by ofedoren**
@ekohl, there is another, more general and detailed approach trying to fix the same issue. I hope the code and the comments will help to understand the idea of what's going wrong.

Unfortunately, that's the last thing I could find. The third and (that's only my opinion) the last option would be to completely get rid of that weird `app/models/host.rb` module, which (per my understanding) simply tries to simulate an abstract class.

UPD: TIL that rubocop can apparently yell that the indentation of a comment can be wrong and that's unacceptable :facepalm: 

Alternative to https://github.com/theforeman/foreman/pull/9871.

### CODE REVIEW

File: app/models/host/base.rb [hunk 1]
```diff
@@ -90,11 +90,12 @@ def self.taxonomy_conditions
     # initializer and we set name when we are sure that we have primary interface
     # we can't create primary interface before calling super because args may contain nested
     # interface attributes
-    def initialize(*args)
+    def initialize(attributes = nil, &block)
```
Comment (ekohl) [inline app/models/host/base.rb]: Looks like we don't override this in discovery (https://github.com/theforeman/foreman_discovery/blob/develop/app/models/host/discovered.rb), which is good.

File: app/models/concerns/foreman/sti.rb [hunk 2]
```diff
@@ -8,11 +8,11 @@ class << base
 
     module ClassMethods
       # ensures that the correct STI object is created when :type is passed.
-      def new(*attributes, &block)
-        if (h = attributes.first).is_a?(Hash) && (type = h.with_indifferent_access.delete(:type)) && !type.empty?
+      def new(attributes = nil, &block)
```
Comment (ekohl) [inline app/models/concerns/foreman/sti.rb]: It took me a while to figure out where it's defined, but `show-source ActiveRecord::Base.new` in pry helped. This is indeed the correct definition, since https://api.rubyonrails.org/classes/ActiveRecord/Inheritance/ClassMethods.html#method-i-new uses the same.

File: test/helpers/application_helper_test.rb
Comment (ekohl) [inline test/helpers/application_helper_test.rb:1]: We can already merge this part, so let's do that. Keeping this PR as small as possible is a good thing.

**Review by ekohl — COMMENTED** [review]
I had a quick look at this and it appears the approach is to make the constructor's signature match the base class. That makes sense to me.

As for STI, I'm not sure we can get rid of it. I'd like to. At this moment I think we only have 2 implementations, so be sure to check out what Discovery does.

File: test/helpers/application_helper_test.rb
Comment (ofedoren) [reply test/helpers/application_helper_test.rb:1]: Extracted to https://github.com/theforeman/foreman/pull/9992.

**Comment by ofedoren** [comment]
Even though initially I wanted to get the (any) fix merged at first, and then take a look at the plugins which fail, but after I've tested with foreman_discovery it seems some additional fixes might be needed in this/other PR as well, so I'll rather start a list here with verified plugins which support this PR:

 - [x] discovery: https://github.com/theforeman/foreman_discovery/pull/622
 - [x] templates: https://github.com/theforeman/foreman_templates/pull/181
 - [x] webhooks: https://github.com/theforeman/foreman_webhooks/pull/72
 - [x] openscap: https://github.com/theforeman/foreman_openscap/pull/557
 - [x] tasks: https://github.com/theforeman/foreman-tasks/pull/743
 - [x] REX: https://github.com/theforeman/foreman_remote_execution/pull/864
 - [x] Katello: https://github.com/Katello/katello/pull/10885
 - [ ] ansible: https://github.com/theforeman/foreman_ansible/pull/681
 - [x] google: https://github.com/theforeman/foreman_google/pull/67
 - [x] puppet: https://github.com/theforeman/foreman_puppet/pull/383
 - [x] bootdisk: https://github.com/theforeman/foreman_bootdisk/pull/156
 - [x] azure: https://github.com/theforeman/foreman_azure_rm/pull/184
 - [x] kubevirt: https://github.com/theforeman/foreman_kubevirt/pull/154
 - [x] leapp: https://github.com/theforeman/foreman_leapp/pull/136
 - [x] virt_who_configure: https://github.com/theforeman/foreman_virt_who_configure/pull/188

File: .github/matrix.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
   "postgresql": ["12"],
-  "ruby": ["2.7"],
+  "ruby": ["2.7", "3.0"],
```
Comment (ekohl) [inline .github/matrix.json:3]: Can we separate this out to its own commit? I'd like to merge all the preparation work before we pull the trigger.

**Review by ekohl — APPROVED** [review]

**Review by ekohl — CHANGES_REQUESTED** [review]
There are some deprecation warnings when running tests:
```
2024-01-23T18:54:00.9931233Z Mocha deprecation warning at /home/runner/work/foreman/foreman/app/services/proxy_api/dhcp.rb:36:in `unused_ip': Expectation defined at /home/runner/work/foreman/foreman/test/unit/proxy_api/dhcp_test.rb:99:in `block in <class:ProxyApiDhcpTest>' expected positional hash ({:query => "mac=00:11:22:33:44:55"}), but received keyword arguments (:query => "mac=00:11:22:33:44:55"). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.
2024-01-23T18:54:01.0281310Z Mocha deprecation warning at /home/runner/work/foreman/foreman/app/services/proxy_api/dhcp.rb:36:in `unused_ip': Expectation defined at /home/runner/work/foreman/foreman/test/unit/proxy_api/dhcp_test.rb:74:in `block in <class:ProxyApiDhcpTest>' expected positional hash ({:query => ""}), but received keyword arguments (:query => ""). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.
2024-01-23T18:54:01.0407752Z Mocha deprecation warning at /home/runner/work/foreman/foreman/app/services/proxy_api/dhcp.rb:36:in `unused_ip': Expectation defined at /home/runner/work/foreman/foreman/test/unit/proxy_api/dhcp_test.rb:90:in `block in <class:ProxyApiDhcpTest>' expected positional hash ({:query => "from=192.168.0.50&to=192.168.0.150"}), but received keyword arguments (:query => "from=192.168.0.50&to=192.168.0.150"). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.
2024-01-23T18:54:13.7290895Z /home/runner/work/foreman/foreman/vendor/bundle/ruby/2.7.0/gems/fog-vsphere-3.6.3/lib/fog/vsphere/compute.rb:344: warning: deprecated Object#=~ is called on Integer; it always returns nil
2024-01-23T18:54:13.7458370Z /home/runner/work/foreman/foreman/vendor/bundle/ruby/2.7.0/gems/fog-vsphere-3.6.3/lib/fog/vsphere/compute.rb:344: warning: deprecated Object#=~ is called on Integer; it always returns nil
2024-01-23T18:54:22.8953091Z Mocha deprecation warning at /home/runner/work/foreman/foreman/app/models/host/managed.rb:715:in `ipmi_boot': Expectation defined at /home/runner/work/foreman/foreman/test/models/host_test.rb:3245:in `block (2 levels) in <class:HostTest>' expected keyword arguments (:function => "bootdevice", :device => "bios"), but received positional hash ({:function => "bootdevice", :device => "bios"}). These will stop matching when strict keyword argument matching is enabled. See the documentation for Mocha::Configuration#strict_keyword_argument_matching=.
```

The fog-vsphere warnings will break with Ruby 3.2. The other ones look like some things we should fix in our code and then enable `strict_keyword_argument_matching` in mocha.

**Comment by ekohl** [comment]
Can we also enable strict keyword argument matching (https://mocha.jamesmead.org/Mocha/Configuration.html#strict_keyword_argument_matching=-instance_method)? Perhaps make that the whole second commit?

**Comment by ofedoren** [comment]
Done. Not sure though if we need that. As there was stated, this will be `true` by default anyway.

**Comment by evgeni** [comment]
> Done. Not sure though if we need that. As there was stated, this will be `true` by default anyway.

That way we don't accidentally regress now that we fixed the offenders until it's `true` in the future?

**Comment by ekohl** [comment]
Indeed. If it was `true` today then these tests wouldn't have failed. This way we also force plugins to adopt.

**Comment by ofedoren** [comment]
Some Katello test failures are related to strict keyword argument matching I guess, but otherwise :green_apple: for Foreman.

**Comment by ekohl** [comment]
> Some Katello test failures are related to strict keyword argument matching I guess, but otherwise 🍏 for Foreman.

I've love to leverage what you did in https://github.com/theforeman/foreman/pull/9989#issuecomment-1889272996 and see how other repositories behave.

**Comment by ofedoren** [comment]
> I've love to leverage what you did in [#9989 (comment)](https://github.com/theforeman/foreman/pull/9989#issuecomment-1889272996) and see how other repositories behave.

I'm preparing more PRs regarding what I found in Katello, I'll paste them here once they are ready. The plan to go over my locally checkedout other plugins as well.

**Review by ekohl — APPROVED** [review]
